R/ZIPLNfit-class.R

```diff
@@ -0,0 +1,831 @@
+      if (!anyNA(Omega))  private$Omega  <- Omega
+      if (!anyNA(Sigma))  private$Sigma  <- Sigma
+      if (!anyNA(M))      private$M      <- M
+      if (!anyNA(S))      private$S      <- S
+      if (!anyNA(R))      private$R      <- R
+      if (!anyNA(Z))      private$Z      <- Z
+      if (!anyNA(A))      private$A      <- A
+      if (!anyNA(Ji))     private$Ji     <- Ji
+      if (!anyNA(monitoring)) private$monitoring <- monitoring
+    },
+
+    #' @description Initialize a [`ZIPLNfit`] model
+    #' @importFrom stats glm.fit residuals poisson fitted coef
+    #' @importFrom pscl zeroinfl
+    initialize = function(responses, covariates, offsets, weights, formula, control) {
+      ## problem dimensions
+      n <- nrow(responses); p <- ncol(responses); d <- ncol(covariates$PLN); d0 <- ncol(covariates$ZI)
+
+      ## save the formula call as specified by the user
+      private$formula <- formula
+      private$X       <- covariates$PLN
+      private$X0      <- covariates$ZI
+      ## initialize the covariance model
+      private$covariance <- control$covariance
+      private$ziparam <- control$ziparam
+
+      R  <- matrix(0, n, p)
+      M  <- matrix(0, n, p)
+      B  <- matrix(0, d , p)
+      B0 <- matrix(0, d0, p)
+      for (j in 1:p) {
+        y = responses[, j]
+        if (min(y) == 0) {
+          suppressWarnings(
+            zip_out <- switch(control$ziparam,
+              "row"    = pscl::zeroinfl(y ~ 0 + covariates$PLN | 0 + factor(1:n), offset = offsets[, j]),
+              "covar"  = pscl::zeroinfl(y ~ 0 + covariates$PLN | 0 + covariates$ZI  , offset = offsets[, j]),
+                         pscl::zeroinfl(y ~ 0 + covariates$PLN |               1, offset = offsets[, j])) # offset only for the count model
+          )
+          B0[,j] <- coef(zip_out, "zero")
+          B[,j]  <- coef(zip_out, "count")
+          R[, j] <- predict(zip_out, type = "zero")
+          M[,j]  <- residuals(zip_out) + covariates$PLN %*% coef(zip_out, "count")
+        } else {
+          p_out <- glm(y ~ 0 + covariates$PLN, family = 'poisson', offset = offsets[, j])
+          B0[,j] <- rep(-10, d)
+          B[,j]  <- coef(p_out)
+          R[, j] <- 0
+          M[,j]  <- residuals(p_out) + covariates$PLN %*% coef(p_out)
+        }
+      }
+
+      ## Initialization of the ZI component
+      private$R  <- R
+      private$Pi <- switch(control$ziparam,
+        "single" = matrix(    mean(R), n, p)              ,
+        "row"    = matrix(rowMeans(R), n, p)              ,
+        "col"    = matrix(colMeans(R), n, p, byrow = TRUE),
+        "covar"  = R)
+      private$B0 <- B0
+      private$zeros <- 1 * (responses == 0)
+
+      ## Initialization of the PLN component
+      private$B <- B
+      private$M <- M
+      private$S <- matrix(.1, n, p)
+
+      ## Link to functions performing the optimization
+      private$optimizer$B     <- function(M, X, Omega, control) optim_zipln_B_dense(M, X)
```

Comment:
`Omega` and `control` are unused arguments for `optim_zipln_B_dense()`. Are they included for forward compatibility (in case we ever need them) ?

Reply:
There was a sparse variant for estimation of B, which I did not included yet, that required `Omega` and the sparsity level, carried by `control`. But they are not needed at this stage, so we can indeed simplify the code.